chacha20/src/backend/avx2.rs

```diff
@@ -17,15 +17,106 @@ use core::arch::x86::*;
 #[cfg(target_arch = "x86_64")]
 use core::arch::x86_64::*;
 
+/// The number of blocks processed per invocation by this backend.
+const BLOCKS: usize = 4;
+
 /// Helper union for accessing per-block state.
 ///
 /// ChaCha20 block state is stored in four 32-bit words, so we can process two blocks in
 /// parallel. We store the state words as a union to enable cheap transformations between
 /// their interpretations.
+///
+/// Additionally, we process four blocks at a time to take advantage of ILP.
 #[derive(Clone, Copy)]
 union StateWord {
-    blocks: [__m128i; 2],
-    avx: __m256i,
+    blocks: [__m128i; BLOCKS],
+    avx: [__m256i; BLOCKS / 2],
+}
+
+impl StateWord {
+    #[inline]
+    #[target_feature(enable = "avx2")]
+    unsafe fn add_assign_epi32(&mut self, rhs: &Self) {
+        self.avx = [
+            _mm256_add_epi32(self.avx[0], rhs.avx[0]),
+            _mm256_add_epi32(self.avx[1], rhs.avx[1]),
+        ];
+    }
+
+    #[inline]
+    #[target_feature(enable = "avx2")]
+    unsafe fn xor_assign(&mut self, rhs: &Self) {
+        self.avx = [
+            _mm256_xor_si256(self.avx[0], rhs.avx[0]),
+            _mm256_xor_si256(self.avx[1], rhs.avx[1]),
+        ];
+    }
+
+    #[inline]
+    #[target_feature(enable = "avx2")]
+    unsafe fn shuffle_epi32<const MASK: i32>(&mut self) {
+        self.avx = [
+            _mm256_shuffle_epi32(self.avx[0], MASK),
+            _mm256_shuffle_epi32(self.avx[1], MASK),
+        ];
+    }
```

Comment:
Bleh, it turns out `_mm256_shuffle_epi32` is a pseudo-function generated such that the second argument is required to be `const`, but there's no way to pass a `const` as a function argument in our own code. This approach therefore requires const generics, which requires a MSRV of 1.51.

Reply:
Yeah, I ended up using const generics like that here:

https://github.com/RustCrypto/universal-hashes/blob/master/polyval/src/backend/pmull.rs#L99-L106

Personally I'd be fine with bumping MSRV to 1.51. It would also let us do a `zeroize` 1.4 bump.